#Data Science

In [350]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC 
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from joblib import dump
from pandas.io.common import infer_compression
from sklearn.ensemble import BaggingRegressor, RandomForestRegressor, RandomForestClassifier

from sklearn.model_selection import KFold, cross_val_score
from sklearn import metrics
from sklearn.model_selection import GridSearchCV

from sklearn.preprocessing import OneHotEncoder

from sklearn import tree
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn import preprocessing 
from sklearn import metrics
from sklearn.metrics import confusion_matrix,accuracy_score


In [351]:
data = pd.read_csv(r"D:\Users\Chaima\Downloads\marketing_campaign1.csv")

#Diviser une colonne en plusieurs colonnes

### PREPROCESSING : CLEANING DATASET


In [352]:
data[['ID','Year_Birth','Education','Marital_Status','Income','Kidhome','Teenhome','Dt_Customer','Recency','MntWines','MntFruits','MntMeatProducts','MntFishProducts','MntSweetProducts','MntGoldProds','NumDealsPurchases','NumWebPurchases','NumCatalogPurchases','NumStorePurchases','NumWebVisitsMonth','AcceptedCmp3','AcceptedCmp4','AcceptedCmp5','AcceptedCmp1','AcceptedCmp2','Complain','Z_CostContact','Z_Revenue','Response']]=data["ID;Year_Birth;Education;Marital_Status;Income;Kidhome;Teenhome;Dt_Customer;Recency;MntWines;MntFruits;MntMeatProducts;MntFishProducts;MntSweetProducts;MntGoldProds;NumDealsPurchases;NumWebPurchases;NumCatalogPurchases;NumStorePurchases;NumWebVisitsMonth;AcceptedCmp3;AcceptedCmp4;AcceptedCmp5;AcceptedCmp1;AcceptedCmp2;Complain;Z_CostContact;Z_Revenue;Response"].str.split(";", expand=True)

#Supprimer la premiére colonne vide

In [353]:
data.columns

Index(['ID;Year_Birth;Education;Marital_Status;Income;Kidhome;Teenhome;Dt_Customer;Recency;MntWines;MntFruits;MntMeatProducts;MntFishProducts;MntSweetProducts;MntGoldProds;NumDealsPurchases;NumWebPurchases;NumCatalogPurchases;NumStorePurchases;NumWebVisitsMonth;AcceptedCmp3;AcceptedCmp4;AcceptedCmp5;AcceptedCmp1;AcceptedCmp2;Complain;Z_CostContact;Z_Revenue;Response',
       'ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Teenhome', 'Dt_Customer', 'Recency', 'MntWines', 'MntFruits',
       'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth',
       'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1',
       'AcceptedCmp2', 'Complain', 'Z_CostContact', 'Z_Revenue', 'Response'],
      dtype='object')

In [354]:
data=data.drop(["ID;Year_Birth;Education;Marital_Status;Income;Kidhome;Teenhome;Dt_Customer;Recency;MntWines;MntFruits;MntMeatProducts;MntFishProducts;MntSweetProducts;MntGoldProds;NumDealsPurchases;NumWebPurchases;NumCatalogPurchases;NumStorePurchases;NumWebVisitsMonth;AcceptedCmp3;AcceptedCmp4;AcceptedCmp5;AcceptedCmp1;AcceptedCmp2;Complain;Z_CostContact;Z_Revenue;Response"], axis=1)

In [355]:
# ID najmou ndroppiwha tawa w mba3d nraj3ou fi ekher etape
#ndroppiw  'Z_Revenue', 'Z_CostContact' 
#bech ndroppiw el revenue ki na3mlou train w fit khater bech na3mlou prediction 3al revenue men kol categorie client 



In [356]:
df= data.copy()

In [357]:
df = df.drop('Z_CostContact',axis=1)

In [358]:
df = df.drop('Z_Revenue',axis=1)

In [359]:
df["Year_Birth"].astype(int)

0       1957
1       1954
2       1965
3       1984
4       1981
        ... 
2235    1967
2236    1946
2237    1981
2238    1956
2239    1954
Name: Year_Birth, Length: 2240, dtype: int32

In [360]:
df['Age'] = 2021- df["Year_Birth"].astype(int)
df = df.drop('Year_Birth',axis=1)
df.head()

ID   Education Marital_Status Income Kidhome Teenhome Dt_Customer  \
0  5524  Graduation         Single  58138       0        0  04/09/2012   
1  2174  Graduation         Single  46344       1        1  08/03/2014   
2  4141  Graduation       Together  71613       0        0  21/08/2013   
3  6182  Graduation       Together  26646       1        0  10/02/2014   
4  5324         PhD        Married  58293       1        0  19/01/2014   

  Recency MntWines MntFruits  ... NumStorePurchases NumWebVisitsMonth  \
0      58      635        88  ...                 4                 7   
1      38       11         1  ...                 2                 5   
2      26      426        49  ...                10                 4   
3      26       11         4  ...                 4                 6   
4      94      173        43  ...                 6                 5   

  AcceptedCmp3 AcceptedCmp4 AcceptedCmp5 AcceptedCmp1 AcceptedCmp2 Complain  \
0            0            0            0            0            0        0   
1            0            0            0            0            0        0   
2            0            0            0            0            0        0   
3            0            0            0            0            0        0   
4            0            0            0            0            0        0   

  Response Age  
0        1  64  
1        0  67  
2        0  56  
3        0  37  
4        0  40  

[5 rows x 27 columns]

In [361]:
#DtCustomer - date of customer’s enrolment with the company

"""
df['Dt_Customer'] = data.to_datetime(df['Dt_Customer'])
df['Dt_Customer_year_month'] = df['Dt_Customer'].dt.to_period("M")
df['account_age'] = (pd.to_datetime("2021-12").year - df['Dt_Customer_year_month'].dt.year)*12 + (pd.to_datetime("2014-12").month - df['Dt_Customer_year_month'].dt.month)
#del df['Dt_Customer_year_month']
#del df['Dt_Customer']
df.head()"""
df = df.drop('Dt_Customer',axis=1)

In [362]:
df.head()

ID   Education Marital_Status Income Kidhome Teenhome Recency MntWines  \
0  5524  Graduation         Single  58138       0        0      58      635   
1  2174  Graduation         Single  46344       1        1      38       11   
2  4141  Graduation       Together  71613       0        0      26      426   
3  6182  Graduation       Together  26646       1        0      26       11   
4  5324         PhD        Married  58293       1        0      94      173   

  MntFruits MntMeatProducts  ... NumStorePurchases NumWebVisitsMonth  \
0        88             546  ...                 4                 7   
1         1               6  ...                 2                 5   
2        49             127  ...                10                 4   
3         4              20  ...                 4                 6   
4        43             118  ...                 6                 5   

  AcceptedCmp3 AcceptedCmp4 AcceptedCmp5 AcceptedCmp1 AcceptedCmp2 Complain  \
0            0            0            0            0            0        0   
1            0            0            0            0            0        0   
2            0            0            0            0            0        0   
3            0            0            0            0            0        0   
4            0            0            0            0            0        0   

  Response Age  
0        1  64  
1        0  67  
2        0  56  
3        0  37  
4        0  40  

[5 rows x 26 columns]

In [408]:
df["Income"] = pd.to_numeric(data["Income"], downcast="float")


In [409]:
# Fillna fel income khatrou ne9es valeurs ma ywarihech 

avg_income = np.mean(df.Income)
df['Income'] = df['Income'].fillna(avg_income, axis=0)

In [410]:
np.mean(df.Income)

52247.254

In [411]:
df.Education.unique()

array([2, 4, 3, 1, 0])

In [412]:
df.Marital_Status.unique()

array([1, 2, 3, 0, 4, 5, 6, 7])

In [413]:
df.Marital_Status = df['Marital_Status'].map({'Single':1, 'Together':2, 'Married':3, 'Divorced':0, 'Widow':4, 'Alone':5,
       'Absurd':6, 'YOLO':7}).astype(int)

IntCastingNaNError: Cannot convert non-finite values (NA or inf) to integer

In [414]:
df.Education= df['Education'].map({'Graduation':2, 'PhD':4, 'Master':3, 'Basic':1, '2n Cycle':0}).astype(int)

IntCastingNaNError: Cannot convert non-finite values (NA or inf) to integer

In [415]:
#CHOIX 1 #ONE HOT ENCODER : 1 W 0 W YFUSIONI M3A DATASET
#CHOIX 2 #NMAPPIW

In [416]:
df.head()

ID  Education  Marital_Status   Income Kidhome Teenhome Recency  \
0  5524          2               1  58138.0       0        0      58   
1  2174          2               1  46344.0       1        1      38   
2  4141          2               2  71613.0       0        0      26   
3  6182          2               2  26646.0       1        0      26   
4  5324          4               3  58293.0       1        0      94   

   MntWines  MntFruits  MntMeatProducts  ...  Age  total_Mnt  MntWines_tot  \
0     635.0       88.0            546.0  ...   64     1617.0      0.392703   
1      11.0        1.0              6.0  ...   67       27.0      0.407407   
2     426.0       49.0            127.0  ...   56      776.0      0.548969   
3      11.0        4.0             20.0  ...   37       53.0      0.207547   
4     173.0       43.0            118.0  ...   40      422.0      0.409953   

   MntFruits_tot  MntMeatProducts_tot  MntFishProducts_tot  \
0       0.054422             0.337662             0.106370   
1       0.037037             0.222222             0.074074   
2       0.063144             0.163660             0.143041   
3       0.075472             0.377358             0.188679   
4       0.101896             0.279621             0.109005   

   MntSweetProducts_tot  MntGoldProds_tot  AcceptedCmps  Total_Purchases  
0              0.054422          0.054422           0.0             32.0  
1              0.037037          0.222222           0.0             11.0  
2              0.027062          0.054124           0.0             25.0  
3              0.056604          0.094340           0.0             14.0  
4              0.063981          0.035545           0.0             24.0  

[5 rows x 35 columns]

In [417]:
"""df['MntWines', 'MntFruits',
       'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds'].astype(float)"""
df["MntWines"] = pd.to_numeric(data["MntWines"], downcast="float")
df["MntFruits"] = pd.to_numeric(df["MntFruits"], downcast="float")
df["MntMeatProducts"] = pd.to_numeric(df["MntMeatProducts"], downcast="float")
df["MntFishProducts"] = pd.to_numeric(df["MntFishProducts"], downcast="float")
df["MntSweetProducts"] = pd.to_numeric(df["MntSweetProducts"], downcast="float")
df["MntGoldProds"] = pd.to_numeric(df["MntGoldProds"], downcast="float")

In [418]:
df["total_Mnt"] = df["MntWines"] + df["MntFruits"] + df["MntMeatProducts"]+ df['MntFishProducts'] + df["MntSweetProducts"] + df["MntGoldProds"]

df['MntWines_tot'] = df['MntWines']/df['total_Mnt']
df['MntFruits_tot'] = df["MntFruits"]/df['total_Mnt']
df["MntMeatProducts_tot"] = df["MntMeatProducts"]/df['total_Mnt']
df["MntFishProducts_tot"] = df["MntFishProducts"]/df['total_Mnt']
df["MntSweetProducts_tot"] = df["MntSweetProducts"]/df['total_Mnt']
df["MntGoldProds_tot"] = df["MntGoldProds"]/df['total_Mnt']

In [419]:
df.describe()

Education  Marital_Status         Income     MntWines    MntFruits  \
count  2240.000000     2240.000000    2240.000000  2240.000000  2240.000000   
mean      2.393750        2.045089   52247.253906   303.935699    26.302233   
std       1.124797        1.094965   25037.796875   336.597412    39.773434   
min       0.000000        0.000000    1730.000000     0.000000     0.000000   
25%       2.000000        1.000000   35538.750000    23.750000     1.000000   
50%       2.000000        2.000000   51741.500000   173.500000     8.000000   
75%       3.000000        3.000000   68289.750000   504.250000    33.000000   
max       4.000000        7.000000  666666.000000  1493.000000   199.000000   

       MntMeatProducts  MntFishProducts  MntSweetProducts  MntGoldProds  \
count      2240.000000      2240.000000       2240.000000   2240.000000   
mean        166.949997        37.525448         27.062946     44.021873   
std         225.715378        54.628979         41.280499     52.167439   
min           0.000000         0.000000          0.000000      0.000000   
25%          16.000000         3.000000          1.000000      9.000000   
50%          67.000000        12.000000          8.000000     24.000000   
75%         232.000000        50.000000         33.000000     56.000000   
max        1725.000000       259.000000        263.000000    362.000000   

       NumDealsPurchases  ...          Age    total_Mnt  MntWines_tot  \
count        2240.000000  ...  2240.000000  2240.000000   2240.000000   
mean            2.325000  ...    52.194196   605.798218      0.458481   
std             1.932238  ...    11.984069   602.249329      0.228411   
min             0.000000  ...    25.000000     5.000000      0.000000   
25%             1.000000  ...    44.000000    68.750000      0.289506   
50%             2.000000  ...    51.000000   396.000000      0.457482   
75%             3.000000  ...    62.000000  1045.500000      0.639143   
max            15.000000  ...   128.000000  2525.000000      0.963303   

       MntFruits_tot  MntMeatProducts_tot  MntFishProducts_tot  \
count    2240.000000          2240.000000          2240.000000   
mean        0.049505             0.249508             0.071532   
std         0.055867             0.126633             0.077909   
min         0.000000             0.000000             0.000000   
25%         0.008924             0.156250             0.012571   
50%         0.029840             0.233447             0.048193   
75%         0.070237             0.328227             0.104703   
max         0.445545             0.997110             0.590909   

       MntSweetProducts_tot  MntGoldProds_tot  AcceptedCmps  Total_Purchases  
count           2240.000000       2240.000000   2240.000000      2240.000000  
mean               0.050777          0.120197      0.297768        20.178572  
std                0.061184          0.108988      0.678381         7.293275  
min                0.000000          0.000000      0.000000         0.000000  
25%                0.008621          0.038095      0.000000        14.000000  
50%                0.033333          0.086294      0.000000        19.000000  
75%                0.070411          0.169768      0.000000        25.000000  
max                0.945848          0.894150      4.000000        46.000000  

[8 rows x 29 columns]

In [420]:
"""ohe= OneHotEncoder(handle_unknown='ignore')
ohe_df = pd.DataFrame(ohe.fit_transform(df[['Education']]).toarray())

df= df.join(ohe_df)

df.head()"""

"ohe= OneHotEncoder(handle_unknown='ignore')\nohe_df = pd.DataFrame(ohe.fit_transform(df[['Education']]).toarray())\n\ndf= df.join(ohe_df)\n\ndf.head()"

In [421]:
df["AcceptedCmp1"] = pd.to_numeric(df["AcceptedCmp1"], downcast="float")
df["AcceptedCmp2"] = pd.to_numeric(df["AcceptedCmp2"], downcast="float")
df["AcceptedCmp3"] = pd.to_numeric(df["AcceptedCmp3"], downcast="float")
df["AcceptedCmp4"] = pd.to_numeric(df["AcceptedCmp4"], downcast="float")
df["AcceptedCmp5"] = pd.to_numeric(df["AcceptedCmp5"], downcast="float")



In [422]:
df["AcceptedCmps"] = df[["AcceptedCmp1","AcceptedCmp2","AcceptedCmp3","AcceptedCmp4", "AcceptedCmp5"]].sum(axis = 1)


In [423]:
df["NumCatalogPurchases"] = pd.to_numeric(df["NumCatalogPurchases"], downcast="float")
df["NumStorePurchases"] = pd.to_numeric(df["NumStorePurchases"], downcast="float")
df["NumWebPurchases"] = pd.to_numeric(df["NumWebPurchases"], downcast="float")
df["NumWebVisitsMonth"] = pd.to_numeric(df["NumWebVisitsMonth"], downcast="float")
df["NumDealsPurchases"] = pd.to_numeric(df["NumDealsPurchases"], downcast="float")

In [424]:
df["Total_Purchases"] = df[["NumCatalogPurchases","NumStorePurchases","NumWebPurchases","NumWebVisitsMonth","NumDealsPurchases"]].sum(axis=1) 

In [425]:
df.head()

ID  Education  Marital_Status   Income Kidhome Teenhome Recency  \
0  5524          2               1  58138.0       0        0      58   
1  2174          2               1  46344.0       1        1      38   
2  4141          2               2  71613.0       0        0      26   
3  6182          2               2  26646.0       1        0      26   
4  5324          4               3  58293.0       1        0      94   

   MntWines  MntFruits  MntMeatProducts  ...  Age  total_Mnt  MntWines_tot  \
0     635.0       88.0            546.0  ...   64     1617.0      0.392703   
1      11.0        1.0              6.0  ...   67       27.0      0.407407   
2     426.0       49.0            127.0  ...   56      776.0      0.548969   
3      11.0        4.0             20.0  ...   37       53.0      0.207547   
4     173.0       43.0            118.0  ...   40      422.0      0.409953   

   MntFruits_tot  MntMeatProducts_tot  MntFishProducts_tot  \
0       0.054422             0.337662             0.106370   
1       0.037037             0.222222             0.074074   
2       0.063144             0.163660             0.143041   
3       0.075472             0.377358             0.188679   
4       0.101896             0.279621             0.109005   

   MntSweetProducts_tot  MntGoldProds_tot  AcceptedCmps  Total_Purchases  
0              0.054422          0.054422           0.0             32.0  
1              0.037037          0.222222           0.0             11.0  
2              0.027062          0.054124           0.0             25.0  
3              0.056604          0.094340           0.0             14.0  
4              0.063981          0.035545           0.0             24.0  

[5 rows x 35 columns]

In [426]:
df_clean = df.copy()

In [427]:
df_clean = df_clean.drop('ID',axis=1)


In [428]:
df_clean = df_clean.drop('Recency',axis=1)

In [429]:
df_clean = df_clean.drop('NumWebVisitsMonth',axis=1)

In [430]:
df_clean = df_clean.drop('Complain',axis=1)

In [431]:
df_clean['Income'] = df_clean['Income'].fillna(avg_income, axis=0)

In [432]:
df_clean['Income'] = df_clean['Income'].str.replace('',avg_income)

AttributeError: Can only use .str accessor with string values!

In [433]:
#["NumCatalogPurchases","NumStorePurchases","NumWebPurchases","NumWebVisitsMonth","NumDealsPurchases"]].sum(axis=1) 
df_clean = df_clean.drop('NumCatalogPurchases',axis=1)
df_clean = df_clean.drop('NumStorePurchases',axis=1)
df_clean = df_clean.drop('NumWebPurchases',axis=1)
df_clean = df_clean.drop('NumDealsPurchases',axis=1)

In [434]:
df_clean.drop(['MntGoldProds','AcceptedCmp3','AcceptedCmp4','AcceptedCmp5','AcceptedCmp1','AcceptedCmp2','MntWines','MntFruits','MntMeatProducts','MntFishProducts','MntSweetProducts'],axis=1)

Education  Marital_Status   Income Kidhome Teenhome Response  Age  \
0             2               1  58138.0       0        0        1   64   
1             2               1  46344.0       1        1        0   67   
2             2               2  71613.0       0        0        0   56   
3             2               2  26646.0       1        0        0   37   
4             4               3  58293.0       1        0        0   40   
...         ...             ...      ...     ...      ...      ...  ...   
2235          2               3  61223.0       0        1        0   54   
2236          4               2  64014.0       2        1        0   75   
2237          2               0  56981.0       0        0        0   40   
2238          3               2  69245.0       0        1        0   65   
2239          4               3  52869.0       1        1        1   67   

      total_Mnt  MntWines_tot  MntFruits_tot  MntMeatProducts_tot  \
0        1617.0      0.392703       0.054422             0.337662   
1          27.0      0.407407       0.037037             0.222222   
2         776.0      0.548969       0.063144             0.163660   
3          53.0      0.207547       0.075472             0.377358   
4         422.0      0.409953       0.101896             0.279621   
...         ...           ...            ...                  ...   
2235     1341.0      0.528710       0.032066             0.135720   
2236      444.0      0.914414       0.000000             0.067568   
2237     1241.0      0.731668       0.038678             0.174859   
2238      843.0      0.507711       0.035587             0.253855   
2239      172.0      0.488372       0.017442             0.354651   

      MntFishProducts_tot  MntSweetProducts_tot  MntGoldProds_tot  \
0                0.106370              0.054422          0.054422   
1                0.074074              0.037037          0.222222   
2                0.143041              0.027062          0.054124   
3                0.188679              0.056604          0.094340   
4                0.109005              0.063981          0.035545   
...                   ...                   ...               ...   
2235             0.031320              0.087994          0.184191   
2236             0.000000              0.000000          0.018018   
2237             0.025786              0.009670          0.019339   
2238             0.094899              0.035587          0.072361   
2239             0.011628              0.005814          0.122093   

      AcceptedCmps  Total_Purchases  
0              0.0             32.0  
1              0.0             11.0  
2              0.0             25.0  
3              0.0             14.0  
4              0.0             24.0  
...            ...              ...  
2235           0.0             23.0  
2236           1.0             29.0  
2237           1.0             25.0  
2238           0.0             26.0  
2239           0.0             18.0  

[2240 rows x 16 columns]

In [435]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 27 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Education             2240 non-null   int32  
 1   Marital_Status        2240 non-null   int32  
 2   Income                2240 non-null   float32
 3   Kidhome               2240 non-null   object 
 4   Teenhome              2240 non-null   object 
 5   MntWines              2240 non-null   float32
 6   MntFruits             2240 non-null   float32
 7   MntMeatProducts       2240 non-null   float32
 8   MntFishProducts       2240 non-null   float32
 9   MntSweetProducts      2240 non-null   float32
 10  MntGoldProds          2240 non-null   float32
 11  AcceptedCmp3          2240 non-null   float32
 12  AcceptedCmp4          2240 non-null   float32
 13  AcceptedCmp5          2240 non-null   float32
 14  AcceptedCmp1          2240 non-null   float32
 15  AcceptedCmp2         

In [436]:
feature = ['Age','Education','Marital_Status','Income','Kidhome','Teenhome','total_Mnt','AcceptedCmps','Total_Purchases']

In [437]:
X=df_clean[feature]

In [438]:
X.sample(10)

Age  Education  Marital_Status   Income Kidhome Teenhome  total_Mnt  \
178    62          2               2  38285.0       2        1       10.0   
1194   52          3               3  43641.0       1        1      139.0   
363    43          4               3  37717.0       1        0       25.0   
1139   60          2               3  49678.0       0        1      315.0   
1728   60          3               2  74881.0       1        1      964.0   
1057   70          4               3  53312.0       0        0      278.0   
1609   57          4               3  65526.0       0        1      506.0   
404    66          4               1  55521.0       1        2      446.0   
839    44          2               1  56628.0       0        1      761.0   
1635   66          4               3  59821.0       0        1      734.0   

      AcceptedCmps  Total_Purchases  
178            0.0              6.0  
1194           0.0             16.0  
363            0.0             13.0  
1139           0.0             21.0  
1728           0.0             32.0  
1057           0.0             19.0  
1609           0.0             19.0  
404            1.0             31.0  
839            1.0             25.0  
1635           0.0             34.0

In [439]:
y = df_clean.Response

In [440]:
df_clean['Income'] = df_clean['Income'].ffill(axis=0) 


In [443]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=10)

In [444]:
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

k_fold = KFold(n_splits = 10, shuffle = True, random_state  =0)

In [445]:
nan_values = df_clean[df_clean.isna().any(axis=1)]
print(nan_values)

Empty DataFrame
Columns: [Education, Marital_Status, Income, Kidhome, Teenhome, MntWines, MntFruits, MntMeatProducts, MntFishProducts, MntSweetProducts, MntGoldProds, AcceptedCmp3, AcceptedCmp4, AcceptedCmp5, AcceptedCmp1, AcceptedCmp2, Response, Age, total_Mnt, MntWines_tot, MntFruits_tot, MntMeatProducts_tot, MntFishProducts_tot, MntSweetProducts_tot, MntGoldProds_tot, AcceptedCmps, Total_Purchases]
Index: []

[0 rows x 27 columns]


In [446]:
rfc = RandomForestClassifier(n_estimators=30,
                             random_state=1)

max_depth_range = range(1,16)
param_grid = dict(max_depth=max_depth_range)

grid = GridSearchCV(rfc,
                    param_grid,
                    cv = 10,
                    scoring = 'accuracy')

grid.fit(X_train, y_train)


grid.cv_results_

{'mean_fit_time': array([0.07338972, 0.05416145, 0.05097506, 0.06259298, 0.11287787,
        0.08310163, 0.10201542, 0.09753826, 0.0849319 , 0.08442893,
        0.0838131 , 0.09378569, 0.11263082, 0.10679941, 0.13208141]),
 'std_fit_time': array([0.03578733, 0.01310019, 0.00748044, 0.01460221, 0.02244506,
        0.02125164, 0.02155873, 0.02648243, 0.01237785, 0.01191774,
        0.00864091, 0.01493563, 0.0135288 , 0.02951447, 0.05057725]),
 'mean_score_time': array([0.00818083, 0.00561304, 0.00533245, 0.00566435, 0.0092329 ,
        0.00708857, 0.0075716 , 0.00852761, 0.00606971, 0.0070066 ,
        0.00668423, 0.00722144, 0.00976634, 0.0085408 , 0.0127203 ]),
 'std_score_time': array([0.00542903, 0.00108153, 0.00143117, 0.00098117, 0.00175903,
        0.00355838, 0.00158282, 0.00283992, 0.00082931, 0.00152068,
        0.00139195, 0.00192988, 0.00332908, 0.00330992, 0.01154164]),
 'param_max_depth': masked_array(data=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
              m

In [447]:
best_rfc = RandomForestClassifier(n_estimators=50,
                                  random_state=1,
                                  max_depth = 12)
best_rfc.fit(X_train, y_train)

rfc_pred = best_rfc.predict(X_test)
accuracy_train = metrics.accuracy_score(y_train, best_rfc.predict(X_train))
accuracy_test = metrics.accuracy_score(y_test, rfc_pred)
print("Accuracy of Random Forest train is: ", accuracy_train)
print("Accuracy of Random Forest test is: ", accuracy_test)

Accuracy of Random Forest train is:  0.9704241071428571
Accuracy of Random Forest test is:  0.90625


In [449]:
forest = RandomForestClassifier()
scoring = 'accuracy'
score = cross_val_score(forest, X, y, cv= k_fold, scoring=scoring)

In [450]:
tree = tree.DecisionTreeClassifier()
scoring = 'accuracy'
score = cross_val_score(tree, X, y, cv= k_fold, n_jobs=1, scoring=scoring)
print(score)

[0.79910714 0.83035714 0.80803571 0.79464286 0.83035714 0.80803571
 0.80803571 0.86160714 0.80357143 0.8125    ]


In [458]:
forest = RandomForestClassifier()
scoring = 'accuracy'
score = cross_val_score(forest, X, y, cv= k_fold, scoring=scoring)
rfc_cv = GridSearchCV(forest, param_grid, cv=10,scoring = 'accuracy')
rfc_cv.fit(X, y)
print(score)

ValueError: Invalid parameter n_neighbors for estimator RandomForestClassifier(). Check the list of available parameters with `estimator.get_params().keys()`.

In [452]:
knn = KNeighborsClassifier(n_neighbors = 5)
scoring = 'accuracy'
score = cross_val_score(knn, X, y, cv= k_fold, n_jobs=1, scoring=scoring)
print(score)

[0.79464286 0.82589286 0.8125     0.87946429 0.85267857 0.83482143
 0.8125     0.82142857 0.83482143 0.87053571]


In [453]:
k_scores = []
for i in range(10,40) :   
    knn = KNeighborsClassifier(n_neighbors = i)
    scoring = 'accuracy'
    score = cross_val_score(knn, X, y, cv = k_fold, scoring = scoring)  
    k_scores.append(np.mean(score))
print(k_scores)

[0.8517857142857143, 0.8486607142857142, 0.8549107142857142, 0.8486607142857142, 0.85, 0.8513392857142857, 0.85, 0.8459821428571429, 0.8428571428571429, 0.8455357142857143, 0.8450892857142858, 0.8455357142857143, 0.8441964285714286, 0.8450892857142858, 0.8446428571428571, 0.8446428571428571, 0.8450892857142858, 0.8450892857142858, 0.846875, 0.8482142857142858, 0.85, 0.85, 0.8482142857142858, 0.8455357142857144, 0.8473214285714287, 0.846875, 0.8482142857142858, 0.846875, 0.8477678571428571, 0.8477678571428571]


In [454]:
from sklearn.model_selection import GridSearchCV

param_grid = {'n_neighbors': np.arange(30, 40)}
knn = KNeighborsClassifier()
knn_cv = GridSearchCV(knn, param_grid, cv=10,scoring = 'accuracy')
knn_cv.fit(X, y)

GridSearchCV(cv=10, estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': array([30, 31, 32, 33, 34, 35, 36, 37, 38, 39])},
             scoring='accuracy')

In [456]:
#modify file path
df_res = pd.DataFrame.from_dict(rfc_cv.cv_results_) #We save the results into a dataframe

### SUBMISSION 

In [ ]:

submission.to_csv("submission.csv", index = False)



```
# Ce texte est au format code
```

#On a 2240 job listing, chaqu'un a 29 informations

#Quels sont les types des données, combien elle reserve de mémoire et s'il y a des valeurs manquantes

# Graduation akther Education, Marital status,...
#Ikharj les valeurs uniques
#Les sommes des colonnes.

#Only numerical values

#Iraja3 les 5 premiers lignes

#Iraja3 les 5 derniers lignes

#Iraja3 ligne aléatoire

#Ly mafihomch valeurs tbadalhom b '-'

In [49]:
data = data.replace('-', float("NaN"))

In [50]:
data.describe()

ID;Year_Birth;Education;Marital_Status;Income;Kidhome;Teenhome;Dt_Customer;Recency;MntWines;MntFruits;MntMeatProducts;MntFishProducts;MntSweetProducts;MntGoldProds;NumDealsPurchases;NumWebPurchases;NumCatalogPurchases;NumStorePurchases;NumWebVisitsMonth;AcceptedCmp3;AcceptedCmp4;AcceptedCmp5;AcceptedCmp1;AcceptedCmp2;Complain;Z_CostContact;Z_Revenue;Response  \
count                                                2240                                                                                                                                                                                                                                                                                                                         
unique                                               2240                                                                                                                                                                                                                                                                                                                         
top     5524;1957;Graduation;Single;58138;0;0;04/09/20...                                                                                                                                                                                                                                                                                                                         
freq                                                    1                                                                                                                                                                                                                                                                                                                         

          ID Year_Birth   Education Marital_Status Income Kidhome Teenhome  \
count   2240       2240        2240           2240   2240    2240     2240   
unique  2240         59           5              8   1975       3        3   
top     5524       1976  Graduation        Married              0        0   
freq       1         89        1127            864     24    1293     1158   

       Dt_Customer Recency  ... NumWebVisitsMonth AcceptedCmp3 AcceptedCmp4  \
count         2240    2240  ...              2240         2240         2240   
unique         663     100  ...                16            2            2   
top     31/08/2012      56  ...                 7            0            0   
freq            12      37  ...               393         2077         2073   

       AcceptedCmp5 AcceptedCmp1 AcceptedCmp2 Complain Z_CostContact  \
count          2240         2240         2240     2240          2240   
unique            2            2            2        2             1   
top               0            0            0        0             3   
freq           2077         2096         2210     2219          2240   

       Z_Revenue Response  
count       2240     2240  
unique         1        2  
top           11        0  
freq        2240     1906  

[4 rows x 30 columns]

In [39]:
print('What are the 2 most and least incomes according to education?\n')



for order in [True, False]:
  income = data.sort_values(by="Education", ascending=order)["Income"][:2].values
  # conencter API avec le résultat intermédiaire income
  education = data.sort_values(by="Education", ascending=order)["Education"][:2].values
  # optimize time complexity w space complexity

  for i in range(2):
    print(f'{income[i]}: {education[i]}')
  
  print('\n')

  print('and in each Marital Status?\n')
# check for duplicates through Income value
for category in data['Marital_Status'].unique():
  subset = data[ data['Marital_Status'] == category]
  
  for order in [True, False]:
    income = subset.sort_values(by="Education", ascending=order)["Income"][:2].values
    education = subset.sort_values(by="Education", ascending=order)["Education"][:2].values

    print(f'{category}:')
    for i in range(2):
      print(f'{income[i]}: {education[i]}')
    print('\n')



What are the 2 most and least income according to education?

47025: 2n Cycle
37284: 2n Cycle


and in each Marital Status?

52869: PhD
37401: PhD


and in each Marital Status?

Single:
89572: 2n Cycle
38593: 2n Cycle


Single:
29999: PhD
66582: PhD


Together:
63855: 2n Cycle
54342: 2n Cycle


Together:
71604: PhD
28764: PhD


Married:
82347: 2n Cycle
27733: 2n Cycle


Married:
58293: PhD
53312: PhD


Divorced:
31163: 2n Cycle
35388: 2n Cycle


Divorced:
25509: PhD
34487: PhD


Widow:
47682: 2n Cycle
28457: 2n Cycle


Widow:
57513: PhD
80995: PhD


Alone:
34176: Graduation
61331: Master


Alone:
35860: PhD
61331: Master


Absurd:
79244: Graduation
65487: Master


Absurd:
65487: Master
79244: Graduation


YOLO:
48432: PhD
48432: PhD


YOLO:
48432: PhD
48432: PhD




#Machine Learning

In [19]:
knn = KNeighborsClassifier()
tree = DecisionTreeClassifier()
svm = SVC()
log = LogisticRegression()
naive = GaussianNB()

In [21]:
data["Kidhome"] = pd.to_numeric(data["Kidhome"], downcast="float")
data["Teenhome"] = pd.to_numeric(data["Teenhome"], downcast="float")
data["NumCatalogPurchases"] = pd.to_numeric(data["NumCatalogPurchases"], downcast="float")
data["NumStorePurchases"] = pd.to_numeric(data["NumStorePurchases"], downcast="float")
features = data[['Kidhome','Teenhome', 'NumCatalogPurchases', 'NumStorePurchases']]
labels = data['NumStorePurchases']

In [22]:
X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size=0.2)


In [24]:
knn.fit(X_train, y_train)
tree.fit(X_train, y_train)
log.fit(X_train, y_train)
naive.fit(X_train, y_train)
svm.fit(X_train, y_train)

C:\Users\Lenovo\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


SVC()

In [28]:
# Option 1: badihya
print(knn.score(X_test, y_test),
tree.score(X_test, y_test),
log.score(X_test, y_test),
naive.score(X_test, y_test),
svm.score(X_test, y_test))
# score = Accuracy

0.984375 1.0 0.8303571428571429 1.0 0.9977678571428571


In [30]:
# Option 2: 

tree_preds = tree.predict(X_test)
#overfitted



In [31]:

confusion_matrix(tree_preds, y_test)

array([[  2,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0],
       [  0,   1,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0],
       [  0,   0,  41,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0],
       [  0,   0,   0, 105,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0],
       [  0,   0,   0,   0,  65,   0,   0,   0,   0,   0,   0,   0,   0,
          0],
       [  0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,   0,   0,
          0],
       [  0,   0,   0,   0,   0,   0,  45,   0,   0,   0,   0,   0,   0,
          0],
       [  0,   0,   0,   0,   0,   0,   0,  25,   0,   0,   0,   0,   0,
          0],
       [  0,   0,   0,   0,   0,   0,   0,   0,  26,   0,   0,   0,   0,
          0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,  15,   0,   0,   0,
          0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  30,   0,   0,
          0],
       [  0,   0,   0,   0,   0,   0,   0, 

In [36]:
sorted(list(zip(tree.feature_importances_, data.columns)), reverse=True)

[(1.0, 'Marital_Status'), (0.0, 'Year_Birth'), (0.0, 'ID'), (0.0, 'Education')]

0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
       ... 
2235    NaN
2236    NaN
2237    NaN
2238    NaN
2239    NaN
Name: NumStorePurchases, Length: 2240, dtype: object